2_process/src/process_nhdv2_attr.R

```diff
@@ -269,3 +269,175 @@ process_catchment_nhdv2_attr <- function(file_path,vars_table,segs_w_comids,nhd_
+  #There must be a typo in the ScienceBase table's units because using m length gives
+  #results that are 3 orders of magnitude larger than other values
+  ind_areas <- mutate(ind_areas, str_dens = length_km/CAT_BASIN_AREA_sum) %>%
+    select(-length_km, -CAT_BASIN_AREA_sum)
+  #assign to attribute table
+  nhdv2_attr_refined <- mutate(nhdv2_attr_refined,
+                          CAT_STRM_DENS_area_wtd = case_when(PRMS_segid %in% ind_areas$PRMS_segid ~
+                                                               ind_areas$str_dens[match(PRMS_segid, ind_areas$PRMS_segid)],
+                                                             TRUE ~ CAT_STRM_DENS_area_wtd)
+  )
+  
+  #Compute TOT variables from PRMS CAT variables
+  # These 3 variables have some segments with 0s that are clearly incorrect:
+  # CWD, TAV7100, TMIN7100 
+  # and TOT for STRM_DENS should be recomputed due to the edits to CAT variables
+  #Add hru segment identifier to match with the recursive function output ID
+  nhdv2_attr_refined$hru_segment <- apply(str_split(nhdv2_attr_refined$PRMS_segid, 
+                                                    pattern = '_', simplify = T), 
+                                          MARGIN = 1, FUN = first) %>%
+    as.numeric()
+  
+  TOT_frmCAT_cols <- c('TOT_CWD_frmCAT', 'TOT_TAV7100_ANN_frmCAT', 
+                       'TOT_TMIN7100_frmCAT', 'TOT_STRM_DENS_frmCAT')
+  nhdv2_attr_refined[TOT_frmCAT_cols] <- NA
+  #loop over all segments
+  for(i in 1:length(unique(nhdv2_attr_refined$hru_segment))){
+    #current segment
+    subseg <- unique(nhdv2_attr_refined$hru_segment)[i]
+    #all upstream segments, including current segment
+    ind_segs <- filter(prms_attribute_df, subseg_seg == subseg) %>% 
+      pull(all_from_segs)
+    #area for each PRMS segment
+    areas_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_BASIN_AREA_sum)
+    #CAT variables for each segment
+    CWD_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_CWD_area_wtd)
+    TAV_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_TAV7100_ANN_area_wtd)
+    TMN_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_TMIN7100_area_wtd)
+    STR_DENS_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_STRM_DENS_area_wtd)
+    
+    #Compute TOT values
+    #need special handling for PRMS segments with _2
+    ind_replace <- which(nhdv2_attr_refined$hru_segment == subseg)
+    nhdv2_attr_refined$TOT_CWD_frmCAT[ind_replace] <- weighted.mean(CWD_segs,areas_segs)
+    nhdv2_attr_refined$TOT_TAV7100_ANN_frmCAT[ind_replace] <- weighted.mean(TAV_segs,areas_segs)
+    nhdv2_attr_refined$TOT_TMIN7100_frmCAT[ind_replace] <- weighted.mean(TMN_segs,areas_segs)
+    nhdv2_attr_refined$TOT_STRM_DENS_frmCAT[ind_replace] <- weighted.mean(STR_DENS_segs,areas_segs)
```

Comment:
Another enhancement suggestion to reduce redundant lines of code. Could you use `purrr::map2()` to loop through two vectors simultaneously (TOT_frmCAT_cols and Cat_variables) and populate the appropriate columns with the right weighted mean values?  

`TOT_frmCAT_cols <- c('TOT_CWD_frmCAT', 'TOT_TAV7100_ANN_frmCAT', 'TOT_TMIN7100_frmCAT', 'TOT_STRM_DENS_frmCAT')`
`Cat_variables <- c(CWD_segs, TAV_segs, TMN_segs, STR_DENS_segs)

Although map2 might not work as needed since we are populating specific columns. Considered providing a suggestion until realizing this.